### PR TITLE
Add watch to mock etcd

### PR DESCRIPTION
**Pull request by mrwwheat7**

### CODE REVIEW

File: src/mock_etcd.rs [hunk 1]
```diff
@@ -76,13 +83,31 @@ impl MockEtcdServer {
...
 struct MockEtcd {
     /// map to store key value
     map: Arc<RwLock<HashMap<Vec<u8>, KeyValue>>>,
+
+    /// map to store key and watch ids
+    watch: Arc<RwLock<HashMap<i64, KeyRange>>>,
+
+    /// map to store watch id and watch response senders
+    #[allow(clippy::type_complexity)]
+    watch_response_sender: Arc<RwLock<HashMap<i64, Arc<RwLock<DuplexSink<WatchResponse>>>>>>,
```
Comment (rogercloud) [inline src/mock_etcd.rs]: We'd better define the type with a name, it's hard to understand for such a complex type.
For example:
``` rust
type LockedDuplexSink = Arc<RwLock<DuplexSink<WatchRespnse>>>;
```

File: src/mock_etcd.rs [hunk 2]
```diff
@@ -128,22 +155,129 @@ impl MockEtcd {
         kvs
     }
 
+    /// Send watch response for a specific watch id to etcd client
+    async fn send_watch_response_with_watch_id(
+        watch_response_sender: Arc<RwLock<DuplexSink<WatchResponse>>>,
```
Comment (rogercloud) [inline src/mock_etcd.rs]: In this context, we know the sender is a watch_response_sender, so we could name it "sender".

File: src/mock_etcd.rs [hunk 3]
```diff
@@ -128,22 +155,129 @@ impl MockEtcd {
...
+            .await
+            .unwrap_or_else(|e| panic!("Fail to send watch response, the error is {}", e));
+    }
+
+    /// Send watch response to etcd client
+    #[allow(clippy::pattern_type_mismatch)]
+    #[allow(clippy::type_complexity)]
+    async fn send_watch_responses(
+        watch_arc: Arc<RwLock<HashMap<i64, KeyRange>>>,
+        watch_response_sender_arc: Arc<
```
Comment (rogercloud) [inline src/mock_etcd.rs]: actually it's not an Arc for the sender, instead it's an Arc for the hashmap. Please rename it.

File: src/mock_etcd.rs (hunk 1 above)
Comment (rogercloud) [reply src/mock_etcd.rs]: If we only have write operation for the DuplexSink, use Mutex instead of a Rwlock.

File: src/mock_etcd.rs [hunk 4]
```diff
@@ -128,22 +155,129 @@ impl MockEtcd {
...
+        >,
+        kv: KeyValue,
+        prev_kv: Option<KeyValue>,
+        event_type: Event_EventType,
+    ) {
+        // Find all watch ids which watch this key and send watch response
+        let watch = watch_arc.read().await;
+        let watch_response_sender = watch_response_sender_arc.read().await;
+        for (watch_id, v) in watch.iter() {
+            let sender = watch_response_sender
```
Comment (rogercloud) [inline src/mock_etcd.rs:201]: This get operation is not efficient for a large watch list for that only a small part or even none of them needs to send a watch response.

File: src/mock_etcd.rs [hunk 5]
```diff
@@ -609,4 +893,147 @@ mod test {
             assert_eq!(resp.count(), 0);
         }));
     }
+
+    fn e2e_watch_test() {
+        let mut etcd_server = MockEtcdServer::new();
+        etcd_server.start();
+
+        smol::future::block_on(Compat::new(async {
```
Comment (rogercloud) [inline src/mock_etcd.rs:894]: How do you know this function is finished when you run it in detach mode?

File: src/mock_etcd.rs [hunk 6]
```diff
@@ -609,4 +893,147 @@ mod test {
...
+            })
+            .detach();
+
+            client
+                .kv()
+                .put(PutRequest::new(key000.clone(), key000.clone()))
+                .await
+                .unwrap_or_else(|err| {
+                    panic!("failed to put key-value key000, the error is {}", err)
+                });
```
Comment (rogercloud) [inline src/mock_etcd.rs]: try to wrap it in a function instead of repeating it.

File: src/mock_etcd.rs (hunk 4 above)
Comment (mrwwheat7) [reply src/mock_etcd.rs:201]: Yes, it is just for test and a very easy solution. If needed, we can implement it by range tree in future.

File: src/mock_etcd.rs (hunk 5 above)
Comment (mrwwheat7) [reply src/mock_etcd.rs:901]: You mean `smol::spawn(async move ` this function? We do not check if this function is finished, we are just checking if the watch responses are correct.

**Review by rogercloud — APPROVED** [review]
